File: backlash.js
const parseFraction = (str) =>
    str.split("/")
        .map(parseFloat)
        .reduce((a, c) => a / c);

class Working {
    static HUES = ["Ruby", "Topaz", "Lemon", "Malachite", "Sapphire", "Ebony", "Over-Ebony"];
    static PITCHES = ["Double-Flat", "Flat", "Natural", "Sharp", "Double-Sharp"];
    static WEAVES = ["Sparse", "Loose", "Tight", "Locked"];
    static LAW_FACTOR = "7/12";
    static DISPLAY_PRECISION = 10**3;
    
    constructor(caspers, hue, pitch, weave, law = Working.LAW_FACTOR) {
        this.caspers = caspers;
        this.hue = hue;
        this.pitch = pitch;
        this.weave = weave;
        this.law = law;
    }
    
    backlash() {
        let newCaspers = this.caspers * parseFraction(this.law);
        
        // hue is inverted, except Over-Ebony, which remains constant
        let newHue = this.hue;
        if(newHue !== "Over-Ebony") {
            let hueIndex = Working.HUES.indexOf(this.hue);
            let newHueIndex = (Working.HUES.length - 2) - hueIndex;
            newHue = Working.HUES[newHueIndex];
        }
        
        // pitch is inverted
        let pitchIndex = Working.PITCHES.indexOf(this.pitch);
        let newPitchIndex = (Working.PITCHES.length - 1) - pitchIndex;
        let newPitch = Working.PITCHES[newPitchIndex];
        
        // weave remains constant
        let newWeave = this.weave;
        
        return new Working(
            newCaspers, newHue, newPitch, newWeave,
            this.law
        );
    }
    
    getDisplayCaspers() {
        let { caspers } = this;
        let [ base, pow ] = caspers.toExponential().split("e").map(parseFloat);
        let suffix;
        if(pow <= 2) {
            suffix = " Caspers";
        }
        else {
            suffix = " Kilocaspers";
            caspers /= 1000;
        }
        // round to 3 max precision
        caspers = Math.round(caspers * Working.DISPLAY_PRECISION) / Working.DISPLAY_PRECISION;
        return caspers + suffix;
    }
    
    getAdjectivalCaspers() {
        return this.getDisplayCaspers().replace(" ", "-").slice(0, -1);
    }
    
    toString() {
        return `${this.caspers}, ${this.hue}, ${this.pitch}, ${this.weavve}`;
    }
    
    getDescriptor() {
        return `${this.getAdjectivalCaspers()} (${this.hue}, ${this.pitch}, ${this.weave})`;
    }
    
    getButton(statButtons) {
        let button = document.createElement("button");
        button.textContent = "Load this";
        // button.textContent = "\u2190 Load this";
        button.title = this.getDescriptor();
        button.addEventListener("click", () => {
            this.loadTo(statButtons);
        });
        return button;
    }
    
    loadTo(statButtons) {
        let [ caspers, hue, pitch, weave, lawFactor ] = statButtons;
        caspers.value = this.caspers;
        hue.value = this.hue;
        pitch.value = this.pitch;
        weave.value = this.weave;
        lawFactor.value = this.law;
    }
}

File: backlash.rb
$HUES = %w(Ruby Topaz Lemon Malachite Sapphire Ebony Over-Ebony)
$PITCHES = %w(Double-Flat Flat Natural Sharp Double-Sharp)
$WEAVE = %w(Sparse Loose Tight Locked)
$LAW_FACTOR = 7.0/12.0


class Working
    attr_reader :caspers, :hue, :pitch, :weave
    
    def initialize(caspers, hue, pitch, weave)
        @caspers = caspers.to_f
        @hue = hue
        @pitch = pitch
        @weave = weave
        # @hue = $HUES.find_index hue
        # @pitch = $PITCHES.find_index pitch
        # @weave = $WEAVE.find_index weave
    end
    
    def backlash
        new_capsers = @caspers * $LAW_FACTOR
        
        new_hue = @hue
        if @hue != "Over-Ebony"
            hue_index = $HUES.find_index @hue
            new_hue_index = ($HUES.size - 2) - hue_index
            new_hue = $HUES[new_hue_index]
        end
        
        pitch_index = $PITCHES.find_index @pitch
        new_pitch_index = ($PITCHES.size - 1) - pitch_index
        new_pitch = $PITCHES[new_pitch_index]
        
        new_weave = @weave
        
        Working.new new_capsers, new_hue, new_pitch, new_weave
    end
    
    def to_s
        "#{@caspers}, #{@hue}, #{@pitch}, #{@weave}"
    end
end

$BACKGROUND = 100

example = Working.new 3000, "Ebony", "Sharp", "Tight"

until example.caspers <= $BACKGROUND
    puts example
    example = example.backlash
end